stl/inc/barrier

```diff
@@ -0,0 +1,192 @@
+// barrier standard header
+
+// Copyright (c) Microsoft Corporation.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#pragma once
+#ifndef _BARRIER_
+#define _BARRIER_
+#include <yvals.h>
+#if _STL_COMPILER_PREPROCESSOR
+
+#ifdef _M_CEE_PURE
+#error <barrier> is not supported when compiling with /clr:pure.
+#endif // _M_CEE_PURE
+
+#if !_HAS_CXX20
+#pragma message("The contents of <barrier> are available only with C++20 or later.")
+#else // ^^^ !_HAS_CXX20 / _HAS_CXX20 vvv
+
+#include <atomic>
+#include <limits.h>
+#include <type_traits>
+#include <xmemory>
+
+#pragma pack(push, _CRT_PACKING)
+#pragma warning(push, _STL_WARNING_LEVEL)
+#pragma warning(disable : _STL_DISABLED_WARNINGS)
+_STL_DISABLE_CLANG_WARNINGS
+#pragma push_macro("new")
+#undef new
+
+_STD_BEGIN
+
+struct _No_completion_function {
+    void operator()() noexcept {}
+};
+
+template <class _Completion_function = _No_completion_function>
+class barrier;
+
+inline constexpr ptrdiff_t _Barrier_arrival_token_mask = 1;
+inline constexpr ptrdiff_t _Barrier_value_mask         = ~_Barrier_arrival_token_mask;
+inline constexpr ptrdiff_t _Barrier_value_shift        = 1;
+inline constexpr ptrdiff_t _Barrier_invalid_token      = 0;
+inline constexpr ptrdiff_t _Barrier_value_step         = 1 << _Barrier_value_shift;
+inline constexpr ptrdiff_t _Barrier_max                = (1ULL << (sizeof(ptrdiff_t) * CHAR_BIT - 2)) - 1;
+
+template <class _Completion_function>
+class _Arrival_token {
+public:
+    _Arrival_token(_Arrival_token&& _Other) noexcept {
+        _Value        = _Other._Value;
+        _Other._Value = _Barrier_invalid_token;
+    }
+
+    _Arrival_token& operator=(_Arrival_token&& _Other) noexcept {
+        _Value        = _Other._Value;
+        _Other._Value = _Barrier_invalid_token;
+    }
+
+private:
+    explicit _Arrival_token(ptrdiff_t _Value_) noexcept : _Value(_Value_) {}
+    friend barrier<_Completion_function>;
+
+    ptrdiff_t _Value;
+};
+
+template <class _Completion_function>
+class barrier {
+public:
+    static_assert(
+#ifndef __cpp_noexcept_function_type

```

Comment:
The standard doesn't allow for this exception; can you motivate why we are allowing it here?

Reply:
not defined `__cpp_noexcept_function_type` in C++20 is a fallback mode.
plus we are not required to diagnose this situation, it is UB to pass throwing functor, not ill-formed.

Reply:
That's fair; it says 'shall be true' which is normally a requirement on users, but it isn't in a mandates block and the user had to ask for a nonstandard mode to get there anyway.